### PR TITLE
Fix stacked areas overflow on null values when connectNulls is true

### LINKED ISSUE #17135: [Bug] stack for null value with connectNulls: true, the color from the upper values overflows to the bottom.
### Version

ng-echarts 8.0.1 

### Link to Minimal Reproduction

https://611udm.csb.app/

### Steps to Reproduce

```

import * as echarts from "echarts";
import "./styles.css";

const chart = echarts.init(document.getElementById("app"));

chart.setOption({
  xAxis: {
    type: "category",
    data: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"]
  },
  yAxis: {
    type: "value"
  },
  series: [
    {
      data: [120, null, 150, 80, 70, 110, 130],
      type: "line",
      connectNulls: true,
      showBackground: true,
      backgroundStyle: {
        color: "rgba(180, 180, 180, 0.2)"
      },
      stack: "test"
    },

    {
      data: [120, null, 150, 80, 70, 110, 130],
      type: "line",
      connectNulls: true,
      areaStyle: {
        color: "red"
      },
      showBackground: true,
      backgroundStyle: {
        color: "rgba(180, 180, 180, 0.2)"
      },
      stack: "test"
    }
  ]
});



### Current Behavior

The area color from the top value overflows to the bottom.

### Expected Behavior

The area color filling stops at the lower value-line for null values

### Environment

```markdown
- OS:
- Browser:
- Framework:
```

### Any additional comments?

_No response_

### LINKED ISSUE #19598: [Bug] `connectNulls` option is ignored by stacked areas
### Version

5.4.3

### Link to Minimal Reproduction

https://echarts.apache.org/examples/en/editor.html?c=line-simple&code=PYBwLglsB2AEC8sDeAoWsAmBDMWDOApmAFzJrqYQC2B0eUdpA2gEQAeLANLCwJ5c9eAJhYBdTuXR5gAVwBOAYwLNJFJLDakAjN17bdQ0gFZYAXwkU1G0kN2loMgDaODpACxmLl5NdgBmO39XWC1PVVFyc3I2AEE2CDxSVEswXhBlWAByBRwCAHNgOV5Mryj0XjiEpNVU9NJMgDcsRxkCTMivQjkIAkTYJlVk71qMzMcIaDavSwUYSYUwADknRz6wOVbpiiw5AiwAZVTHDKQy71pZjBPVS00eDi3LPUEuG7DvWDxcBQBrUhYvlhfq9LGd0EMUmlRuNJiU3rNoPMlis1hsCI9YDs9odeMckmCKBdgFdqh90Hd2CCyc8-CIMQSpN8_jxAcCtqZyBFTABuIA

### Steps to Reproduce

Is visible in the reproduction

### Current Behavior

Screenshot 2024-02-06 at 10 48 26 AM

The green area of the second series (y2) dips below the first series (y) at x=2, where y is null

### Expected Behavior

Screenshot 2024-02-06 at 10 49 47 AM

The green area of the second series (y2) should not dip below the first series (y) at x=2, where y is null. Instead, because the `connectNulls` option is `true`, it should behave as if an interpolated value of y=2 is there, and render its line and area accordingly.

### Environment

```markdown
- OS: macOS 13.6.2
- Browser: Firefox 123.0b7
```

### Any additional comments?

Seems like this issue was reported twice before, but those threads are old and inactive
https://github.com/apache/echarts/issues/12404
https://github.com/apache/echarts/issues/17135

### PR DESCRIPTION
**Pull request by alxnddr**
## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [ ] new feature
- [ ] others

### What does this PR do?

[Minimal reproduction](https://echarts.apache.org/examples/en/editor.html?c=line-simple&code=PYBwLglsB2AEC8sDeAoWsAmBDMWDOApmAFzJrqYQC2B0eUdpA2gEQAeLANLCwJ5c9eAJgF8AzCwC6ncujzAArgCcAxgWayKSWG1IBGbr1JDDQ0gFZDY47AC-Milp3HDpaAoA2H028_fYvNaw7l52Do7IzrBirtE-sAYBQUJhmpLk9uRsAIJsEHikqI5gvCDqsADkKjgEAObASrwV4ZnovLn5hZolZaQVAG5YHgoEFRnhhEoQBAWwTJpFET3lFR4Q0KPhjiowGypgAHJ-s2BKI1sUWEoEWADKJR7lSK0RtDsYT5qOujwcF45GQRcL6pCKwPC4FQAa1ILAhWGhwMcL3Qi2KpRWaw2zRBO2ge0Ox1Ip3OIKuN3uvEehRRFDewA-XTB6B-7CRzMBfBE_1BEXh0Nh_Kh7PQtLRFGWfSxm1xuwI-yOXhOZwIPPJdweT1p6HpjLIzKibJ5bVhgRFFG14MhMJ4QvNtnI6VsAG4gA)
[Fix on the preview branch](https://echarts.apache.org/examples/en/editor.html?version=PR-19602%40536da2c&code=PYBwLglsB2AEC8sDeAoWsAmBDMWDOApmAFzJrqYQC2B0eUdpA2gEQAeLANLCwJ5c9eAJgF8AzCwC6ncujzAArgCcAxgWayKSWG1IBGbr1JDDQ0gFZDY47AC-Milp3HDpaAoA2H028_fYvNaw7l52Do7IzrBirtE-sAYBQUJhmpLk9uRsAIJsEHikqI5gvCDqsADkKjgEAObASrwV4ZnovLn5hZolZaQVAG5YHgoEFRnhhEoQBAWwTJpFET3lFR4Q0KPhjiowGypgAHJ-s2BKI1sUWEoEWADKJR7lSK0RtDsYT5qOujwcF45GQRcL6pCKwPC4FQAa1ILAhWGhwMcL3Qi2KpRWaw2zRBO2ge0Ox1Ip3OIKuN3uvEehRRFDewA-XTB6B-7CRzMBfBE_1BEXh0Nh_Kh7PQtLRFGWfSxm1xuwI-yOXhOZwIPPJdweT1p6HpjLIzKibJ5bVhgRFFG14MhMJ4QvNtnI6VsAG4gA)

The pull request fixes the logic of computing stacked on points of stacked areas which fixes areas overflow on null values when `connectNulls` is true.

Stacked on value can be `NaN` when:
1) When the value of series exists but there is no series below it or the series it is stacked on do not have suitable values according to `stackStrategy`
2) When the value of series itself is `NaN`

The `getStackedOnPoint` function behaved the same in both scenarios, when the stacked on value is `NaN` it decided that the area should be stacked on the `valueStart` of the series coordinates. However, when the value of series itself is `NaN` the stacked on value should be NaN as well for `connectNulls` option to work correctly and just connect the line without dropping it to the start of coordinates.

### Fixed issues

Closes https://github.com/apache/echarts/issues/19598
Closes https://github.com/apache/echarts/issues/17135

## Details

There is an existing test case `area-stack.html` which has a chart affected by the issue.

### Before: What was the problem?

Yellow and blue areas have enabled `connectNulls` so the lines in the middle are connected but areas drop to the start of coordinates.

Screenshot 2024-02-08 at 4 29 21 PM

### After: How does it behave after the fixing?

After the fix not only the lines are connected but also areas.

Screenshot 2024-02-08 at 4 28 39 PM

## Document Info

One of the following should be checked.

- [x] This PR doesn't relate to document changes
- [ ] The document should be updated later
- [ ] The document changes have been made in apache/echarts-doc#xxx

### ZRender Changes

- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

`area-stack.html -> line break` 

### Merging options

- [ ] Please squash the commits into a single one when merging.

### CODE REVIEW

**Comment by helgasoft** [comment]
The purpose of stacked lines is to show totals. This PR would make totals at **all red points incorrect**.
I think responsibility for such "fixes" belongs to the developer, not to the charting library.
Solution is simple - **preprocess your data** by removing or approximating missing values, then ECharts will gladly display what you want to see.

image

**Comment by alxnddr** [comment]
@helgasoft thanks for your feedback. In my pull request I addressed areas stacking when `connectNulls` is enabled so that it literally connects polygon lines on null values instead of dropping the area polygon to zero which does not look correct:
Screenshot 2024-02-08 at 4 29 21 PM

Manually interpolating values would break the `step` property behavior, and in addition to it supporting `smooth` would be a disproportionate effort to implement with custom interpolation.

To summarize
- the existing behavior is not correct which was was reported by a few people 
- the PR changes make `connectNulls` work as it is called — connect line and polygon on null values
- there is no workaround that does not limit users from using other features of ECharts

Based on these points, I believe this change is valid to be merged to the library code.

**Comment by helgasoft** [comment]
> addressed areas stacking when connectNulls is enabled 

I think _connectNulls_ should not be enabled in stacked area charts since it makes ECharts display fake totals.

Here are the two methods of data preprocessing - approximate or remove missing values, [Demo Code](https://echarts.apache.org/examples/en/editor.html?c=line-simple&code=G4QwTgBAHgglCWBnAIiALiCBeCBtAugNwBQokAJuiAIzZ5GngSUYBMdBJZzVAzBw24sQAFgFcmwgKzjGFKgDZZAMwD2kABTd4dAAyEIOgDwRq-wwGoLASggBvYhGhwkqDADoADgFdEACw0AckBvH0A7t0CIC0NrEidhai9fAI0AWgBZdD93MBAAO3JVAFsNWxSIXXdWa3c0VQAxeCgAU3INXmsYxx42RP8NDQy0LJz8opLI8vd2mvrGlraO2O6QXl7kweG8guLbKIqqmYbm1vbOuKoRNf70zOytsdLJg9qj-dOl6SvUjbvRnYgyvtqi85idFl1hAovj8RttxntKsDZscFp0AL7EAD0mIgiCaaAghSQiHguQA5hBQAAbbxNRDEeK4KT4bCBFKBAxOBlUVhMllYNkc7kYXh81nskjCERigUS4UgKQywWSxRKuWYgBUWJxYCahVUwCahOJpIp1Np9Pi7kQnip8AAxk0NFIADTUTrCVjW20Op2u90qkXeu2O51uj0XYO-sMB-VSKOh_0RjBQm0hv3hkiwBAoKgJjOxjWY7UQdQQECeTxgVQIQroI1ExAk8mUkA0uny2g4K11zwaABy3kKACMmmBkyB2N2ee5ewOh6Px4GVnRhKs54OR2OJ2JpxhLhuF9vlzI9wrZxX51ul_KlGeoYfr51bzKIJvF_0NJDcCIWVFvwo-C2DizzIm84LSK-75jp-kG_hMkGAcBECga8YI7lBR5gLBFw_n-yzSkhEAgUiaGosuorMnQ0HYf0a54QhfC4ERJGHKC5HyryVE4DROFsAx_48sxQHEShpHse8naYdefE0AJyzUMJyGoRJ4JyIYaB6gAymgACeVJGjgdgQLk6h1lSABc9hdHauRNDp-lNFZxkAO7wOQQxWewaIul04BNCADkGc5aIQBiaIkOpqieGg8CqLkdAOE4Blkk0-TOQZyhoFZgQKLoACkgQ-RAsVoMF9iaVA2UQIEFZVjW8B1ppEBqJAYRSEVvlOLUqhUsONbOV0TjKAFaDeLqg1OFNhIgGSDoACq6Z4TnWdN016ctVm4IEtlNIELo1cO4D7TViAYPaADWJ2BLFBnkIE-BDVNGLTS9YVdSVqi9bFniTVNaBgPAZKpWAOUgDm-1PZ4qgkrF8VWco3i5PacMJRo0OkmgthJWtupjWACW4BjuRoLguj4AdgRmIVDCvR9Tjg0gAAKqiY2Of3rUtK07aSe1PW9PldNmSAc8sVnC7mGD0xA_VI5QYC6QA4hWCNtni0tpvAaAwLqICi04_iqC5VkA7S_NdILTi6S4iDOU4lvLAAWl9hRbcZG3c6SJLkHtYUQBTXR4oDdJbU9OPTbkICFNzu2Q2tJVczlsfSwb50XWDVJUnHa2ILpI69Vp8AAF4rWYKdiwp5da9penldXhRBU05eIPqqieSVYC0uXjv2vFdko4Ome2x3Zt02HT1OJH0dJ7zrDZ5zm01cnE-4mnGdZ83ef9VShcl1ZZcr8IVmelXmkN7XK31435fYr3uT92gg9UsPptNyvLdfe3r_89L4dTdiU9uYWDvg_J-ttX7z3-onJevNIGpxAJddecDcRbwLsXUuuhy5H2WLwU-NdHJWSvhfcuVIQCjksvYXEfgjYm07kaB201b59yaAPbwQ9aFd3fq3L-dCf7j3jgAqO3MUggJYY_Nhz8LLKDVntcuHsZ52WQWdBB6captg3u_VBO90H70wYfKgx8Lh4PPgQjS-CDIkLIU0ChxlDbGxHvQ8uojWHsJajI5u3C_AcKaHw6af8nCCOnjVERzCXGSIgXI6BPNFHNzXmozOSitG7wwVggxyxXQryIaYrJFiV6kPIXbOx3j3oryYffMRYDvEeM_l4hxvipr-OIpiQBOVgGhPEew6Rz9ZEr3kTAmJ784m1QSZvfO2i96mD0fHbBkJjGN0IWfa-eSrE2KoTQhxJT47OI6ZItx3Tqlt1qd_V6XRHoRSAA)
The developer would choose which one is more appropriate, or a combination of the two.

image
image

**Review by Ovilia — COMMENTED** [review]
While I do agree with @helgasoft that stacked null values make the sum value misleading, I personally think this PR provides a much better solution than our current implementation. And the viewer can judge from the broken area to see that you can't take y values as sums, which I think should be better if the developers don't know how or don't bother to remove null values. Let's see what others think about this feature.

**Comment by helgasoft** [comment]
A strong argument in favor would be to find out whether another charting library (highcharts, amcharts, ?) has taken a chance on implementing it.

**Comment by visakadev** [comment]
> > addressed areas stacking when connectNulls is enabled
> 
> I think _connectNulls_ should not be enabled in stacked area charts since it makes ECharts display fake totals.
> 
> Here are the two methods of data preprocessing - approximate or remove missing values, [Demo Code](https://echarts.apache.org/examples/en/editor.html?c=line-simple&code=G4QwTgBAHgglCWBnAIiALiCBeCBtAugNwBQokAJuiAIzZ5GngSUYBMdBJZzVAzBw24sQAFgFcmwgKzjGFKgDZZAMwD2kABTd4dAAyEIOgDwRq-wwGoLASggBvYhGhwkqDADoADgFdEACw0AckBvH0A7t0CIC0NrEidhai9fAI0AWgBZdD93MBAAO3JVAFsNWxSIXXdWa3c0VQAxeCgAU3INXmsYxx42RP8NDQy0LJz8opLI8vd2mvrGlraO2O6QXl7kweG8guLbKIqqmYbm1vbOuKoRNf70zOytsdLJg9qj-dOl6SvUjbvRnYgyvtqi85idFl1hAovj8RttxntKsDZscFp0AL7EAD0mIgiCaaAghSQiHguQA5hBQAAbbxNRDEeK4KT4bCBFKBAxOBlUVhMllYNkc7kYXh81nskjCERigUS4UgKQywWSxRKuWYgBUWJxYCahVUwCahOJpIp1Np9Pi7kQnip8AAxk0NFIADTUTrCVjW20Op2u90qkXeu2O51uj0XYO-sMB-VSKOh_0RjBQm0hv3hkiwBAoKgJjOxjWY7UQdQQECeTxgVQIQroI1ExAk8mUkA0uny2g4K11zwaABy3kKACMmmBkyB2N2ee5ewOh6Px4GVnRhKs54OR2OJ2JpxhLhuF9vlzI9wrZxX51ul_KlGeoYfr51bzKIJvF_0NJDcCIWVFvwo-C2DizzIm84LSK-75jp-kG_hMkGAcBECga8YI7lBR5gLBFw_n-yzSkhEAgUiaGosuorMnQ0HYf0a54QhfC4ERJGHKC5HyryVE4DROFsAx_48sxQHEShpHse8naYdefE0AJyzUMJyGoRJ4JyIYaB6gAymgACeVJGjgdgQLk6h1lSABc9hdHauRNDp-lNFZxkAO7wOQQxWewaIul04BNCADkGc5aIQBiaIkOpqieGg8CqLkdAOE4Blkk0-TOQZyhoFZgQKLoACkgQ-RAsVoMF9iaVA2UQIEFZVjW8B1ppEBqJAYRSEVvlOLUqhUsONbOV0TjKAFaDeLqg1OFNhIgGSDoACq6Z4TnWdN016ctVm4IEtlNIELo1cO4D7TViAYPaADWJ2BLFBnkIE-BDVNGLTS9YVdSVqi9bFniTVNaBgPAZKpWAOUgDm-1PZ4qgkrF8VWco3i5PacMJRo0OkmgthJWtupjWACW4BjuRoLguj4AdgRmIVDCvR9Tjg0gAAKqiY2Of3rUtK07aSe1PW9PldNmSAc8sVnC7mGD0xA_VI5QYC6QA4hWCNtni0tpvAaAwLqICi04_iqC5VkA7S_NdILTi6S4iDOU4lvLAAWl9hRbcZG3c6SJLkHtYUQBTXR4oDdJbU9OPTbkICFNzu2Q2tJVczlsfSwb50XWDVJUnHa2ILpI69Vp8AAF4rWYKdiwp5da9penldXhRBU05eIPqqieSVYC0uXjv2vFdko4Ome2x3Zt02HT1OJH0dJ7zrDZ5zm01cnE-4mnGdZ83ef9VShcl1ZZcr8IVmelXmkN7XK31435fYr3uT92gg9UsPptNyvLdfe3r_89L4dTdiU9uYWDvg_J-ttX7z3-onJevNIGpxAJddecDcRbwLsXUuuhy5H2WLwU-NdHJWSvhfcuVIQCjksvYXEfgjYm07kaB201b59yaAPbwQ9aFd3fq3L-dCf7j3jgAqO3MUggJYY_Nhz8LLKDVntcuHsZ52WQWdBB6captg3u_VBO90H70wYfKgx8Lh4PPgQjS-CDIkLIU0ChxlDbGxHvQ8uojWHsJajI5u3C_AcKaHw6af8nCCOnjVERzCXGSIgXI6BPNFHNzXmozOSitG7wwVggxyxXQryIaYrJFiV6kPIXbOx3j3oryYffMRYDvEeM_l4hxvipr-OIpiQBOVgGhPEew6Rz9ZEr3kTAmJ784m1QSZvfO2i96mD0fHbBkJjGN0IWfa-eSrE2KoTQhxJT47OI6ZItx3Tqlt1qd_V6XRHoRSAA) The developer would choose which one is more appropriate, or a combination of the two.
> 
> image image

### approximate solution fixed my problem: 💯 
The data preprocessing - **approximate solution fixed** my problem: 

In my case, when rendering the area range (minimum value, maximum value), and latest, I used the solution shown in the code before and applied a function to approximate the values.

Before: 👎 
image

**After approximate null values** 👍 
image

I added a custom tooltip to render the original value instead of the approximate value.
image
image

image
`function replaceNullsWithApproximate(data) {
    let startIndex = null;
    let approximateValue = null;

    for (let i = 0; i < data.length; i++) {
      if (data[i] !== null && !isNaN(parseFloat(data[i]))) {
        const value = parseFloat(data[i]);
        if (startIndex !== null && !isNaN(approximateValue)) {
          const endValue = value;
          const slope = (endValue - approximateValue) / (i - startIndex);
          for (let j = startIndex + 1; j < i; j++) {
            const approximate = approximateValue + slope * (j - startIndex);
            data[j] = Number(approximate.toFixed(3));
          }
        }
        startIndex = i;
        approximateValue = value;
      }
    }

    return data;
  }`

**Comment by admsev** [comment]
@gooroodev review

**Comment by gooroodev** [comment]
### 1. Summary of Changes
The pull request addresses an issue where stacked areas overflow on null values when the `connectNulls` option is set to true. The changes involve updating the `getStackedOnPoint` function in `src/chart/line/helper.ts` to handle null values more gracefully by:
- Separating the retrieval of `stackedOverValue` and `stackResultValue`.
- Adding a condition to check for `stackedOverValue` and `stackResultValue` to determine if `stackedOverValue` should be set to `dataCoordInfo.valueStart`.

### 2. Issues, Bugs, or Typos

#### Issue 1: Unnecessary Check on `dataCoordInfo.stacked`
The condition `dataCoordInfo.stacked && isNaN(stackResultValue)` is redundant since `dataCoordInfo.stacked` is already checked before retrieving `stackResultValue`.

**Proposed Improvement:**
Remove the redundant `dataCoordInfo.stacked` check in the condition.

```diff
-    if (isNaN(stackedOverValue) && !(dataCoordInfo.stacked && isNaN(stackResultValue))) {
+    if (isNaN(stackedOverValue) && !isNaN(stackResultValue)) {
```

### 3. General Review of Code Quality and Style

#### Positive Aspects:
- The code changes are clear and well-targeted at solving the specific issue.
- Variable names are meaningful and improve readability.

#### Suggestions for Improvement:
- Consistency in formatting: Ensure consistent use of indentation and spacing.
- Comments: Adding comments to explain the logic, especially the condition handling, would improve maintainability.

**Example of Improved Code with Comments:**
```typescript
export function getStackedOnPoint(
    dataCoordInfo: DataCoordInfo,
    coordSys: CoordinateSystem,
    data: SeriesData,
    idx: number
) {
    let stackedOverValue = NaN;
    let stackResultValue = NaN;

    // Retrieve stacked values if stacking is enabled
    if (dataCoordInfo.stacked) {
        stackedOverValue = data.get(
            data.getCalculationInfo('stackedOverDimension'),
            idx
        ) as number;
        stackResultValue = data.get(
            data.getCalculationInfo('stackResultDimension'),
            idx
        ) as number;
    }

    // If stackedOverValue is NaN but stackResultValue is valid, use valueStart
    if (isNaN(stackedOverValue) && !isNaN(stackResultValue)) {
        stackedOverValue = dataCoordInfo.valueStart;
    }

    const baseDataOffset = dataCoordInfo.baseDataOffset;
    const stackedData = [];
    stackedData[baseDataOffset] = data.get(dataCoordInfo.baseDim, idx);
    stackedData[1 - baseDataOffset] = stackedOverValue;

    return coordSys.dataToPoint(stackedData);
}
```

Overall, the pull request effectively addresses the issue, but minor improvements in condition handling and code comments would enhance its quality.

Yours, [Gooroo.dev](https://gooroo.dev). To receive reviews automatically, [install Github App](https://github.com/apps/gooroo-dev)
